pkg/logql/log/pipeline_test.go

```diff
@@ -29,27 +29,100 @@ func TestPipeline(t *testing.T) {
 		NewStringLabelFilter(labels.MustNewMatcher(labels.MatchEqual, "foo", "bar")),
 		newMustLineFormatter("lbs {{.foo}}"),
 	})
-	l, lbr, ok := p.ForStream(lbs).Process([]byte("line"))
+	l, lbr, ok := p.ForStream(lbs).Process(0, []byte("line"))
 	require.Equal(t, []byte("lbs bar"), l)
 	require.Equal(t, NewLabelsResult(lbs, lbs.Hash()), lbr)
 	require.Equal(t, true, ok)
 
-	ls, lbr, ok := p.ForStream(lbs).ProcessString("line")
+	ls, lbr, ok := p.ForStream(lbs).ProcessString(0, "line")
 	require.Equal(t, "lbs bar", ls)
 	require.Equal(t, NewLabelsResult(lbs, lbs.Hash()), lbr)
 	require.Equal(t, true, ok)
 
-	l, lbr, ok = p.ForStream(labels.Labels{}).Process([]byte("line"))
+	l, lbr, ok = p.ForStream(labels.Labels{}).Process(0, []byte("line"))
 	require.Equal(t, []byte(nil), l)
 	require.Equal(t, nil, lbr)
 	require.Equal(t, false, ok)
 
-	ls, lbr, ok = p.ForStream(labels.Labels{}).ProcessString("line")
+	ls, lbr, ok = p.ForStream(labels.Labels{}).ProcessString(0, "line")
 	require.Equal(t, "", ls)
 	require.Equal(t, nil, lbr)
 	require.Equal(t, false, ok)
 }
 
+func TestFilteringPipeline(t *testing.T) {
+	p := NewFilteringPipeline([]PipelineFilter{
+		newPipelineFilter(2, 4, labels.Labels{{Name: "foo", Value: "bar"}, {Name: "bar", Value: "baz"}}, "e"),
+		newPipelineFilter(3, 5, labels.Labels{{Name: "baz", Value: "foo"}}, "e"),
+	}, newStubPipeline())
+
+	tt := []struct {
+		name   string
+		ts     int64
+		line   string
+		labels labels.Labels
+		ok     bool
+	}{
+		{"it doesn't fall in the timerange", 1, "line", labels.Labels{{Name: "baz", Value: "foo"}}, true},
+		{"it doesn't match the filter", 3, "all good", labels.Labels{{Name: "baz", Value: "foo"}}, true},
+		{"it doesn't match all the selectors", 3, "line", labels.Labels{{Name: "foo", Value: "bar"}}, true},
+		{"it matches all selectors", 3, "line", labels.Labels{{Name: "foo", Value: "bar"}, {Name: "bar", Value: "baz"}}, false},
+		{"it tries all the filters", 5, "line", labels.Labels{{Name: "baz", Value: "foo"}}, false},
+	}
+
+	for _, test := range tt {
+		t.Run(test.name, func(t *testing.T) {
+			_, _, ok := p.ForStream(test.labels).Process(test.ts, []byte(test.line))
+			require.Equal(t, test.ok, ok)
+
+			_, _, ok = p.ForStream(test.labels).ProcessString(test.ts, test.line)
+			require.Equal(t, test.ok, ok)
+		})
+	}
+}
+
+//nolint:unparam
+func newPipelineFilter(start, end int64, lbls labels.Labels, filter string) PipelineFilter {
+	var stages []Stage
+	var matchers []*labels.Matcher
+	for _, l := range lbls {
+		m := labels.MustNewMatcher(labels.MatchEqual, l.Name, l.Value)
+		stages = append(stages, NewStringLabelFilter(m))
```

Comment:
I think we have to filter on stream. If we select stream `{foo="bar", bar="baz"}` and we have deletes for `{bar="baz"} |= error` then we have to make sure the stream we're filtering matches `bar="baz"` before performing the rest of the filter.

Reply:
We have to match label and filter to do the pipeline filtering. If the delete expression only has label selectors, I think we get label-filtering for free. Is there more to it?

Reply:
Label filtering happens in a pipeline only if it is defined using pipe operator i.e `|`. The labels that you are passing in `ForStream` are source stream labels and `PipelineFilter.Matchers` are the matchers from the delete requests. Instead of hand building pipeline you can do `ParseLogSelector` to avoid confusion.

Reply:
I can't use `ParseLogSelector` because it creates a circular dependency but I made [this gist](https://gist.github.com/MasslessParticle/97d0a236bd92e8cfc22cee58e6c9a819) that can be dropped in the `syntax package` 

I understand some more. I was talking about the wrong thing when I mentioned label filtering, sorry. This is not doing label filtering

Because we don't know which streams will be queried when we get the deletes, there are possibly many deletes per request. Because a given pipeline can We have to make sure the stream matches the delete at runtime time before we apply the pipeline stages. 

For example, we query for `{foo="bar"}` and there are deletes for `{foo="bar"} | "e"` and `{bar="baz"}`. There will be two pipeline filters but we only want to apply `{foo="bar"}`

Pipelines can also be used to filter multiple streams. If the query is `{foo!="ba*"}`, we could pull back multiple streams, but only want to filter `{foo="bar"}` so we can't know at query time which deletes to apply.

When `ForStream` is called, it figures out which deletes to apply and adds those to the list. Any label matching is hidden in those pipelines.